elm/web/rhub.py

```diff
@@ -0,0 +1,646 @@
+"""
+Code to build Corpus from the researcher hub.
+"""
+import os
+import os.path
+import logging
+from urllib.request import urlopen
+import requests
+import pandas as pd
+from bs4 import BeautifulSoup
+
+logger = logging.getLogger(__name__)
+
+
+class ResearchOutputs():
+    """Class to handle publications portion of the NREL researcher hub."""
+    BASE_URL = "https://research-hub.nrel.gov/en/publications/?page=0"
+
+    def __init__(self, url, n_pages=1, txt_dir='./ew_txt'):
+        """
+        Parameters
+        ----------
+        url : str
+            Research hub publications URL, most likely
+            https://research-hub.nrel.gov/en/publications/
+        n_pages : int
+            Number of pages to get from the API. Typical response has 50
+            entries per page. Default of 1 ensures that this class doesnt hang
+            on a million responses.
+        txt_dir : str
+            File directory where you would like to save output .txt files.
+        """
+
+        self.text_dir = txt_dir
+        self.all_links = []
+        for p in range(0, n_pages):
+            url = url + f"?page={p}"
+            html = self.html_response(url)
+            self.soup = BeautifulSoup(html, "html.parser")
+
+            self.target = self.soup.find('ul', {'class': 'list-results'})
+            self.docs = self.target.find_all('a', {'class': 'link'})
+
+            page_links = [d['href'] for d in self.docs if
+                          '/publications/' in d['href']]
+            self.all_links.extend(page_links)
+
+    def html_response(self, url):
+        """Function to retrieve html response.
+
+        Parameters
+        ----------
+        url : str
+            URL of interest.
+
+        Returns
+        -------
+        html : str
+            HTML response output.
+        """
+        with urlopen(url) as page:
+            html = page.read().decode("utf-8")
+
+        return html
+
+    def _scrape_authors(self, soup_inst):
+        """Scrape the names of authors associated with given publication.
+
+        Parameters
+        ----------
+        soup_inst : bs4.BeautifulSoup
+            Active beautiful soup instance for the url associated with a
+            given publication.
+
+        Returns
+        -------
+        authors : list
+            List of all authors (strings) that contributed to publication.
+        """
+
+        authors = soup_inst.find('p', {'class': 'relations persons'}).text
+
+        return authors
+
+    def _scrape_links(self, soup_inst):
+        """Scrape the links under 'Access to Document' header
+        for a publication.
+
+        Parameters
+        ----------
+        soup_inst : bs4.BeautifulSoup
+            Active beautiful soup instance for the url associated with a
+            given publication.
+
+        Returns
+        -------
+        doi link (str)
+        pdf link (str): link to pdf if it exists
```

Comment:
There was a comment that the returns doc string should have a different format. That comment applied to all Returns docstrings. Please update everywhere to this format:
```
doi : str
    Description of doi